lmdeploy/pytorch/backends/default/rotary_embedding.py

```diff
@@ -0,0 +1,354 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+
+import torch
+from torch import nn
+
+from ..rotary_embedding import (Llama3Parameters, LongRoPEScalingParameters,
+                                RopeType, RotaryEmbeddingBuilder,
+                                RotaryEmbeddingImpl, YarnParameters)
+
+
+def _rotary_embedding_fwd(position_ids: torch.Tensor,
+                          inv_freq: torch.Tensor,
+                          scaling_factor: float,
+                          mscale: float = None,
+                          dtype: torch.dtype = None,
+                          device_type: torch.device = None):
+    """rotary embedding forward."""
+    if dtype is None:
+        dtype = torch.float16
+    if device_type is None:
+        device_type = 'cuda'
+    position_ids = position_ids.float() / scaling_factor
+    inv_freq_expanded = inv_freq[None, :,
+                                 None].float().expand(position_ids.shape[0],
+                                                      -1, 1)
+    position_ids_expanded = position_ids[:, None, :]
+    # Force float32 since bfloat16 loses precision on long contexts
+    # See https://github.com/huggingface/transformers/pull/29285
+    device_type = device_type if isinstance(
+        device_type, str) and device_type != 'mps' else 'cpu'
+    with torch.autocast(device_type=device_type, enabled=False):
+        freqs = (inv_freq_expanded.float()
+                 @ position_ids_expanded.float()).transpose(1, 2)
+        emb = freqs.repeat(1, 1, 2)
+        cos = emb.cos()
+        sin = emb.sin()
+
+        if mscale is not None:
+            cos = cos * mscale
+            sin = sin * mscale
+
+    return cos.to(dtype=dtype), sin.to(dtype=dtype)
+
+
+class RotaryEmbeddingImpl(RotaryEmbeddingImpl, nn.Module):
```

Comment:
Could we rename this class or father class? has same name as father class `RotaryEmbeddingImpl`